File: webhook/webhook.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"strconv"
)

// Constants for AES decryption
const (
	AESKey = "oFcTbSzdmOzXF2DZ72h9ILdLsXTE8qK1"
	IV     = "IVa7b9f8d6e54gha"
)

// DecryptPayload function to decrypt the payload
func DecryptPayload(encryptedData string) (string, error) {
	ciphertext, _ := base64.StdEncoding.DecodeString(encryptedData)
	block, err := aes.NewCipher([]byte(AESKey))
	if err != nil {
		return "", err
	}
	if len(ciphertext) < aes.BlockSize {
		return "", fmt.Errorf("ciphertext too short")
	}
	if len(ciphertext)%aes.BlockSize != 0 {
		return "", fmt.Errorf("ciphertext is not a multiple of the block size")
	}

	mode := cipher.NewCBCDecrypter(block, []byte(IV))
	mode.CryptBlocks(ciphertext, ciphertext)

	// PKCS7 Unpadding
	pad := ciphertext[len(ciphertext)-1]
	if int(pad) > len(ciphertext) {
		return "", fmt.Errorf("padding size error")
	}
	unpadLen := len(ciphertext) - int(pad)
	return string(ciphertext[:unpadLen]), nil
}

func verifySignature(payload, receivedSignature, projectId, timestamp string) bool {
	dataToHash := projectId + AESKey + timestamp + payload
	fmt.Println("Data to hash:", dataToHash)
	hash := sha256.Sum256([]byte(dataToHash))
	computedSignature := hex.EncodeToString(hash[:])
	return computedSignature == receivedSignature
}

// WebhookHandler handles incoming webhook requests
func WebhookHandler(w http.ResponseWriter, r *http.Request) {
	body, err := ioutil.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "Error reading request body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	var data map[string]interface{}
	err = json.Unmarshal(body, &data)
	if err != nil {
		http.Error(w, "Error parsing JSON", http.StatusBadRequest)
		return
	}

	fmt.Println("Received webhook:", data)
	decryptedMessage, err := DecryptPayload(data["payload"].(string))
	if err != nil {
		http.Error(w, "Failed to decrypt payload: "+err.Error(), http.StatusInternalServerError)
		return
	}
	fmt.Println("Decrypted payload:", decryptedMessage)

	timestampFloat, _ := data["timestamp"].(float64)
	timestamp := int(timestampFloat)
	timestampStr := strconv.Itoa(timestamp)
	if !verifySignature(decryptedMessage, data["signature"].(string), data["project_id"].(string), timestampStr) {
		fmt.Println("Invalid signature")
		http.Error(w, "Invalid signature", http.StatusForbidden)
		return
	}
	// Add your signature verification logic here

	fmt.Fprintf(w, "ok")
}

func main() {
	http.HandleFunc("/webhook", WebhookHandler)
	log.Fatal(http.ListenAndServe(":3000", nil))
}

File: webhook/webhook.java
import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.Base64;
import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import javax.xml.bind.DatatypeConverter;
import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import org.json.JSONObject;

public class webhook {

    private static final String KEY = "oFcTbSzdmOzXF2DZ72h9ILdLsXTE8qK1";
    private static final String IV = "IVa7b9f8d6e54gha"; // Ensure this is 16 bytes

    public static void main(String[] args) throws Exception {
        int port = 3000;
        ServerSocket serverSocket = new ServerSocket(port);
        System.out.println("Server is listening on port " + port);

        while (true) {
            Socket socket = serverSocket.accept();
            new Thread(new ClientHandler(socket)).start();
        }
    }

    private static class ClientHandler implements Runnable {
        private final Socket socket;

        ClientHandler(Socket socket) {
            this.socket = socket;
        }

        public void run() {
          try {
              InputStream input = socket.getInputStream();
              BufferedReader reader = new BufferedReader(new InputStreamReader(input));

              StringBuilder requestHeaders = new StringBuilder();
              String line;
              while (!(line = reader.readLine()).isEmpty()) {
                  requestHeaders.append(line + "\n");
              }

              // Read the payload (body) after headers
              StringBuilder payloadBuilder = new StringBuilder();
              while(reader.ready()) {
                  payloadBuilder.append((char) reader.read());
              }

              String receiveBody = payloadBuilder.toString().trim(); // Get payload, trim for safety
              System.out.println("Received webhook:" + receiveBody); // Debug print to check payload content

              if (receiveBody.isEmpty()) {
                  System.out.println("No payload found.");
                  return;
              }

              //String decodedBody = URLDecoder.decode(receiveBody, "UTF-8");  // Decode URL-encoded data
              JSONObject postData = new JSONObject(receiveBody);
              String payload = postData.getString("payload");
              //String payload = postData.getString("payload").replaceAll("\\s", "");  // Remove any spaces

              System.out.println("payload: " + payload);
              String decryptedMessage = decrypt(payload);
              System.out.println("Decrypted message: " + decryptedMessage);

              if (!verifySignature(decryptedMessage, postData.getString("signature"), postData.getString("project_id"), String.valueOf(postData.getInt("timestamp")))) {
                System.out.println("Invalid signature");

                OutputStream output = socket.getOutputStream();
                PrintWriter writer = new PrintWriter(output, true);
                writer.println("HTTP/1.1 403 Forbidden");
                writer.println("Content-Type: text/plain");
                writer.println("Connection: close");
                writer.println();
                writer.println("Invalid signature");
                socket.close();
                return;
              }

              OutputStream output = socket.getOutputStream();
              PrintWriter writer = new PrintWriter(output, true);
              writer.println("HTTP/1.1 200 OK\r\n");
              writer.println("Content-Type: text/plain\r\n");
              writer.println("Connection: close\r\n");
              writer.println("\r\n");
              writer.println("ok");

              socket.close();
          } catch (Exception e) {
              e.printStackTrace();
          }
        }

        private static boolean verifySignature(String payload, String receivedSignature, String project_id, String timestamp) {
            try {
                MessageDigest digest = MessageDigest.getInstance("SHA-256");
                byte[] hashedBytes = digest.digest(( project_id + KEY + timestamp + payload).getBytes("UTF-8"));
                String computedSignature = DatatypeConverter.printHexBinary(hashedBytes).toLowerCase();

                return computedSignature.equals(receivedSignature);
            } catch (NoSuchAlgorithmException | UnsupportedEncodingException ex) {
                System.err.println("Error computing digest: " + ex.getMessage());
                return false;
            }
        }

        private String decrypt(String payload) throws Exception {
            IvParameterSpec iv = new IvParameterSpec(IV.getBytes("UTF-8"));
            SecretKeySpec skeySpec = new SecretKeySpec(KEY.getBytes("UTF-8"), "AES");

            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5PADDING");
            cipher.init(Cipher.DECRYPT_MODE, skeySpec, iv);
            byte[] original = cipher.doFinal(Base64.getDecoder().decode(payload));

            return new String(original);
        }
    }

    // Method to extract payload from HTTP request
    private static String parsePayload(String request) {
        // Simplified extraction logic, should parse the request body properly
        return request.substring(request.indexOf("payload=") + 8);
    }
}

